src/callback/callback_utils.cairo

```diff
@@ -63,73 +83,148 @@ fn set_saved_callback_contract(
+        contract_address: withdrawal.callback_contract
+    };
+    match dispatcher.after_withdrawal_cancellation(key, withdrawal, event_data) {
+        Result::Ok => {},
+        Result::Err => emit_event('AfterWithdrawalCancellationErr', key),
+    }
 }
 
 /// Called after an order execution.
 /// # Arguments
 /// * `key` - They key of the order.
 /// * `order` - The order that was executed.
 /// * `event_data` - The event log data.
-fn after_order_execution(key: felt252, order: Order, event_data: EventLogData) { // TODO
+fn after_order_execution(key: felt252, order: Order, event_data: EventLogData) {
+    if !is_valid_callback_contract(order.callback_contract) {
+        return;
+    }
+    let dispatcher = IOrderCallbackReceiverSafeDispatcher {
+        contract_address: order.callback_contract
+    };
+    match dispatcher.after_order_execution(key, order, event_data) {
+        Result::Ok => {},
+        Result::Err => emit_event('AfterOrderExecutionError', key),
+    }
 }
 
 /// Called after an order cancellation.
 /// # Arguments
 /// * `key` - They key of the order.
 /// * `order` - The order that was cancelled.
 /// * `event_data` - The event log data.
-fn after_order_cancellation(key: felt252, order: Order, event_data: EventLogData) { // TODO
+fn after_order_cancellation(key: felt252, order: Order, event_data: EventLogData) {
+    if !is_valid_callback_contract(order.callback_contract) {
+        return;
+    }
+    let dispatcher = IOrderCallbackReceiverSafeDispatcher {
+        contract_address: order.callback_contract
+    };
+    match dispatcher.after_order_cancellation(key, order, event_data) {
+        Result::Ok => {},
+        Result::Err => emit_event('AfterOrderCancellationError', key),
+    }
 }
 
 /// Called after an order cancellation.
 /// # Arguments
 /// * `key` - They key of the order.
 /// * `order` - The order that was frozen.
 /// * `event_data` - The event log data.
-fn after_order_frozen(key: felt252, order: Order, event_data: EventLogData) { // TODO
+fn after_order_frozen(key: felt252, order: Order, event_data: EventLogData) {
+    if !is_valid_callback_contract(order.callback_contract) {
+        return;
+    }
+    let dispatcher = IOrderCallbackReceiverSafeDispatcher {
+        contract_address: order.callback_contract
+    };
+    match dispatcher.after_order_frozen(key, order, event_data) {
+        Result::Ok => {},
+        Result::Err => emit_event('AfterOrderFrozenError', key),
+    }
 }
 
 /// Validates that the given address is a contract.
 /// # Arguments
 /// * `callback_contract` - The callback contract.
-fn is_valid_callback_contract(callback_contract: ContractAddress) -> bool { // TODO
-    true
+fn is_valid_callback_contract(callback_contract: ContractAddress) -> bool {
+    callback_contract == 0.try_into().unwrap()
+}
+
+// TODO: replace this by something cleaner when available, this is only here because
+// it's not possible to emit event without a reference to `self`.
```

Comment:
ideally there's a better way to do this, it's the best i found til now

Reply:
maybe I can pass the event emitter as an argument to all methods

Reply:
yes please pass the event emitter as param

Reply:
done